src/op_mode/dhcp.py

```diff
@@ -0,0 +1,274 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2022 VyOS maintainers and contributors
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 2 or later as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import sys
+from ipaddress import ip_address
+import typing
+
+from datetime import datetime
+from sys import exit
+from tabulate import tabulate
+from isc_dhcp_leases import IscDhcpLeases
+
+from vyos.base import Warning
+from vyos.configquery import ConfigTreeQuery
+
+from vyos.util import cmd
+from vyos.util import dict_search
+from vyos.util import is_systemd_service_running
+
+import vyos.opmode
+
+
+config = ConfigTreeQuery()
+pool_key = "shared-networkname"
+
+
+def _in_pool(lease, pool):
+    if pool_key in lease.sets:
+        if lease.sets[pool_key] == pool:
+            return True
+    return False
+
+
+def _utc_to_local(utc_dt):
+    return datetime.fromtimestamp((utc_dt - datetime(1970, 1, 1)).total_seconds())
+
+
+def _format_hex_string(in_str):
+    out_str = ""
+    # if input is divisible by 2, add : every 2 chars
+    if len(in_str) > 0 and len(in_str) % 2 == 0:
+        out_str = ':'.join(a+b for a,b in zip(in_str[::2], in_str[1::2]))
+    else:
+        out_str = in_str
+
+    return out_str
+
+
+def _find_list_of_dict_index(lst, key='ip', value='') -> int:
+    """
+    Find the index entry of list of dict matching the dict value
+    Exampe:
+        % lst = [{'ip': '192.0.2.1'}, {'ip': '192.0.2.2'}]
+        % _find_list_of_dict_index(lst, key='ip', value='192.0.2.2')
+        % 1
+    """
+    idx = next((index for (index, d) in enumerate(lst) if d[key] == value), None)
+    return idx
+
+
+def _get_raw_server_leases(family, pool=None) -> list:
+    """
+    Get DHCP server leases
+    :return list
+    """
+    lease_file = '/config/dhcpdv6.leases' if family == 'inet6' else '/config/dhcpd.leases'
+    data = []
+    leases = IscDhcpLeases(lease_file).get()
+    if pool is not None:
+        if config.exists(f'service dhcp-server shared-network-name {pool}'):
+            leases = list(filter(lambda x: _in_pool(x, pool), leases))
+    for lease in leases:
+        data_lease = {}
+        data_lease['ip'] = lease.ip
+        data_lease['state'] = lease.binding_state
+        data_lease['pool'] = lease.sets.get('shared-networkname', '')
+        data_lease['end'] = _utc_to_local(lease.end).strftime('%Y/%m/%d %H:%M:%S')
```

Comment:
I think we should only do timestamp to human-readable date conversion in human-readable output, but raw outputs should use timestamps.
If UIs and scripts want to do their own datetime display or calculation, sending timestamps in the reply will make their life a lot easier.

Reply:
I have an issue with the raw format and timestamp datetime
```
        #data_lease['end'] = _utc_to_local(lease.end).strftime('%Y/%m/%d %H:%M:%S')
        data_lease['end'] = datetime.strptime(str(lease.end), '%Y-%m-%d %H:%M:%S').timestamp()
```
Output:
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet
IP Address    Hardware address    State    Lease start          Lease expiration    Remaining    Pool    Hostname
------------  ------------------  -------  -------------------  ------------------  -----------  ------  ----------
192.0.2.47    ca:4f:54:b9:9a:27   active   2022-10-28 06:25:18  1.66701e+09         21:35:44     Lan01   r1
192.0.2.48    6a:2a:78:2c:6e:54   active   2022-10-28 06:25:18  1.66701e+09         21:35:44     Lan01   r1
192.0.2.49    4a:03:eb:1e:57:d5   active   2022-10-28 06:25:19  1.66701e+09         21:35:45     Lan01   r1
192.0.2.46    9a:07:79:19:72:b3   active   2022-10-28 06:25:21  1.66701e+09         21:35:47     Lan01   r1
192.0.2.45    fa:ca:03:00:35:a7   active   2022-10-28 06:25:22  1.66701e+09         21:35:48     Lan01   r1
vyos@r14:~$ 
```

Raw:
```
vyos@r14:~$ /usr/libexec/vyos/op_mode/dhcp.py show_server_leases --family inet --raw
Traceback (most recent call last):
  File "/usr/libexec/vyos/op_mode/dhcp.py", line 276, in <module>
    res = vyos.opmode.run(sys.modules[__name__])
  File "/usr/lib/python3/dist-packages/vyos/opmode.py", line 152, in run
    return dumps(res, indent=4)
  File "/usr/lib/python3.9/json/__init__.py", line 234, in dumps
    return cls(
  File "/usr/lib/python3.9/json/encoder.py", line 201, in encode
    chunks = list(chunks)
  File "/usr/lib/python3.9/json/encoder.py", line 429, in _iterencode
    yield from _iterencode_list(o, _current_indent_level)
  File "/usr/lib/python3.9/json/encoder.py", line 325, in _iterencode_list
    yield from chunks
  File "/usr/lib/python3.9/json/encoder.py", line 405, in _iterencode_dict
    yield from chunks
  File "/usr/lib/python3.9/json/encoder.py", line 438, in _iterencode
    o = _default(o)
  File "/usr/lib/python3.9/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type datetime is not JSON serializable
vyos@r14:~$ 
```

Reply:
in Progress...

Reply:
Done!